### ISSUE TITLE
Links, inside snippets, contain site's navigation name on 404 pages

### ISSUE THREAD

**Issue by dolmit-tanel-paaro** [pimcore/pimcore#8283]
# Bug Report

### Expected behavior
Links, rendered inside snippets, do not contain site's navigation name.

### Actual behavior
Links, rendered inside snippets, contain site's navigation name.

### Steps to reproduce
1. Admin: create a page (e.g. `test`);
2. Admin: use newly created page as site (e.g. set main domain to be `test.loc`);
3. Admin: create a new page as a child of your site (e.g. `page`);
4. Admin: create a new snippet as a child of your site and set it's template to be `test/snippet.html.twig`;
5. Admin: create a property for your site, e.g. `testSnippet` - set it's value to point towards your snippet (e.g. /test/snippet);
6. Create a new controller under `AppBundle\Controller` namespace, e.g. `TestController` with `testAction` method;
7. Create a template file `app/Resources/view/test/test.html.twig` with contents of:
```
<p>
    {{ pimcore_inc(document.property('testSnippet')) }}
</p>
```
8. Create a template file `app/Resources/view/test/snippet.html.twig` with the contents of:
```
<p>
    {{ pimcore_link('snippetLink') }}
</p>
```
9. Admin: open the child page and set it to use `TestController` and `test` method;
10. Admin: open the snippet in the admin interface and set it's link to point towards the test page (e.g. `/test/page`);
11. Enable production mode (`PIMCORE_ENVIRONMENT=prod`);
12. Admin: open your site's page, open edit tab - **IMPORTANT** to do this first because this generates the cache;
13. Admin: edit your site and set it's error page to point towards the test page (e.g. `/test/page`);
14. Produce a 404 page in the frontend (e.g. `http://test.loc/awd`);
15. Rendered link contains site's navigation name (e.g. `http://test.loc/test/page`). It should be like this: `http://test.loc/page`.

This seems to be a cache conflict - same cache is used for admin and frontend.
For some reason the cache isn't disabled for the snippets rendered by the admin request.

Tested this behavior on version 6.8.10 in my local Pimcore Demo Docker environment.

**Comment by NiklasBr** [comment]
We are also seeing this on 6.8.10

**Comment by brusch** [comment]
Fixed by #8328

**Comment by NiklasBr** [comment]
We have begun experiencing this in 6.9.4 again, it can be fixed for a few hours by clearing the full page cache and visiting the front end in another browser.

**Comment by dvesh3** [comment]
@NiklasBr sorry, not able to reproduce on 6.9.4.

**Comment by dolmit-tanel-paaro** [comment]
@brusch Hello, there seems to be the same issue regarding the admin preview.
We have included this patch (Pimcore 6.9.6) to avoid caching URLs rendered in admin preview:
```diff
--- a/lib/Templating/Renderer/IncludeRenderer.php	2021-06-23 08:26:54.000000000 +0000
+++ b/lib/Templating/Renderer/IncludeRenderer.php	2023-02-15 11:27:09.052545342 +0000
@@ -16,6 +16,7 @@
 namespace Pimcore\Templating\Renderer;
 
 use Pimcore\Cache;
+use Pimcore\Http\RequestHelper;
 use Pimcore\Model;
 use Pimcore\Model\Document\PageSnippet;
 use Pimcore\Model\Document\Targeting\TargetingDocumentInterface;
@@ -36,12 +37,19 @@
      */
     private $targetingConfigurator;
 
+    /**
+     * @var RequestHelper
+     */
+    private $requestHelper;
+
     public function __construct(
         ActionRenderer $actionRenderer,
-        DocumentTargetingConfigurator $targetingConfigurator
+        DocumentTargetingConfigurator $targetingConfigurator,
+        RequestHelper $requestHelper
     ) {
         $this->actionRenderer = $actionRenderer;
         $this->targetingConfigurator = $targetingConfigurator;
+        $this->requestHelper = $requestHelper;
     }
 
     /**
@@ -89,8 +97,9 @@
         // check if output-cache is enabled, if so, we're also using the cache here
         $cacheKey = null;
         $cacheConfig = false;
+        $adminFrontendRequest = $this->requestHelper->isFrontendRequestByAdmin();
 
-        if ($cacheEnabled && !$editmode) {
+        if ($cacheEnabled && !$editmode && !$adminFrontendRequest) {
             if ($cacheConfig = Frontend::isOutputCacheEnabled()) {
                 // cleanup params to avoid serializing Element\ElementInterface objects
                 $cacheParams = $params;
@@ -142,7 +151,7 @@
         \Pimcore\Cache\Runtime::set('pimcore_editmode', $editmodeBackup);
 
         // write contents to the cache, if output-cache is enabled & not in editmode
-        if ($cacheConfig && !$editmode && !DeviceDetector::getInstance()->wasUsed()) {
+        if ($cacheConfig && !$editmode && !DeviceDetector::getInstance()->wasUsed() && !$adminFrontendRequest) {
             Cache::save($content, $cacheKey, ['output', 'output_inline'], $cacheConfig['lifetime']);
         }
```

Please let us know if this is a valid solution.

**Comment by NiklasBr** [comment]
Thanks @dolmit-tanel-paaro we no longer use 6.9 but from reading the code and my memory it does look like it might have helped us fix the remaining issue. In situations like these it would be great if the community was allowed to do PR and issue some patch releases for the previous version of Pimcore.
